add content for FileSystemSyncAccessHandle (#21815)
mdn/content: files/en-us/web/api/filesystemsyncaccesshandle/close/index.md

---
title: FileSystemSyncAccessHandle.close()
slug: Web/API/FileSystemSyncAccessHandle/close
page-type: web-api-instance-method
tags:
  - close
  - Directory
  - File
  - File System Access API
  - FileSystemSyncAccessHandle
  - Method
  - stream
  - working with files
browser-compat: api.FileSystemSyncAccessHandle.close
---

{{securecontext_header}}{{DefaultAPISidebar("File System Access API")}}

The **`close()`** method of the
{{domxref("FileSystemSyncAccessHandle")}} interface closes an open synchronous file handle, disabling any further operations on it and releasing the exclusive lock previously put on the file associated with the file handle.

## Syntax

```js-nolint
close()
```

### Parameters

None.

### Return value

> A {{jsxref('Promise')}} which resolves to undefined.

Reviewer: ```suggestion
undefined
```

Author: According to the spec, and the current async versions, it is a Promise. I've also mentioned the experimental sync versions. When the sync versions become the standard, then we can always change it.

### Exceptions

None.

## Examples

The following asynchronous event handler function is contained inside a Web Worker. On receiving a message from the main thread it:

- Creates a synchronous file access handle.
- Gets the size of the file and creates an {{jsxref("ArrayBuffer")}} to contain it.
- Reads the file contents into the buffer.
- Encodes the message and writes it to the end of the file.
- Persists the changes to disk and closes the access handle.

```js
onmessage = async (e) => {

>   // retrieve message sent to work from main script

Reviewer 2: ```suggestion
  // Retrieve message sent to work from main script
```

  const message = e.data;

  // Get handle to draft file
  const root = await navigator.storage.getDirectory();

>   const draftFile = await root.getFileHandle('draft.txt');
>   // Get sync access handle
>   const accessHandle = await draftFile.createSyncAccessHandle();
> 
>   // Get size of the file.
>   const fileSize = await accessHandle.getSize();
>   // Read file content to a buffer.
>   const buffer = new DataView(new ArrayBuffer(fileSize));
>   const readBuffer = accessHandle.read(buffer, { "at": 0 });
> 
>   // Write the message to the end of the file.
>   const encoder = new TextEncoder();
>   const encodedMessage = encoder.encode(message);
>   const writeBuffer = accessHandle.write(encodedMessage, { "at" : readBuffer });
> 
>   // Persist changes to disk.
>   await accessHandle.flush();
> 
>   // Always close FileSystemSyncAccessHandle if done.
>   await accessHandle.close();

Reviewer: ```suggestion
  const draftHandle = await root.getFileHandle('draft.txt', { create: true });
  // Get sync access handle
  const accessHandle = await draftHandle.createSyncAccessHandle();

  // Get size of the file.
  const fileSize = accessHandle.getSize();
  // Read file content to a buffer.
  const buffer = new DataView(new ArrayBuffer(fileSize));
  const readBuffer = accessHandle.read(buffer, { "at": 0 });

  // Write the message to the end of the file.
  const encoder = new TextEncoder();
  const encodedMessage = encoder.encode(message);
  const writeBuffer = accessHandle.write(encodedMessage, { "at" : readBuffer });

  // Persist changes to disk.
  accessHandle.flush();

  // Always close FileSystemSyncAccessHandle if done.
  accessHandle.close();
```

}
```

> Note that some browsers feature an experimental synchronous version of `close()` that provides further improved performance:

Reviewer: Since https://github.com/whatwg/fs/pull/55 is coming, not sure if you want to keep this.

Author: I'd prefer to change this when the spec is changed over, AND browsers actually implement the sync versions by default.
